### ISSUE TITLE
NForce Connection Issue

### ISSUE THREAD

**Issue by scottmcconnell5** [kevinohara80/nforce#158]
I'm having issues connecting to Salesforce through my app. My goal is to create an account for the user when a button is clicked, and that the account is created by automatically logging into Salesforce using my own username/password/security token. My app is written in javascript/html/css and was created using Meteor. Here is the relevant code (**all capitalized variables I have hardcoded with actual values**):

```
case "SalesForceButton":
    var nforce = require('nforce');

    var org = nforce.createConnection({
        clientId: 'CLIENT_ID',
        clientSecret: 'CLIENT_SECRET',
        redirectUri: 'https://localhost:3000/oauth/_callback',
        environment: 'production',
        mode: 'single'
    });
    var username      = 'USERNAME',
    password      = 'PASSWORD',
    securityToken = 'SECURITY_TOKEN',
    oauth;

    org.authenticate({ username: username, password: password, securityToken: securityToken }, 
    function(err, resp){
        if(!err) {
            console.log('Access Token: ' + resp.access_token);
            oauth = resp;

            var acc = nforce.createSObject('Account');
            acc.set('Name', 'Spiffy Cleaners');
            acc.set('Phone', '800-555-2345');
            org.insert({sobject: acc}, function(err, resp){
                if(!err) console.log('It worked!');
            });
        } else {
            console.log('Error: ' + err.message);
        }

    });
break;
```
Things I've done:
1) Setup nforce properly (npm install nforce)
2) Created a connected app on Salesforce using the same redirect URI as above and https://localhost:3000 as the start URL
3) Whitelisted https://localhost:3000 in the CORS section in Salesforce
4) Put https://localhost:3000 in Remote Site Settings section in Salesforce
5) Done all the steps above but using a publicly hosted URL (https://www.my_company.com)
6) Used the "Access-Control-Allow-Origin" chrome extension (it shortened the second error below but console was still showing "Response to preflight request doesn't pass access control check: Response had HTTP status code 400."

Errors I'm getting:
![screen shot 2018-07-31 at 12 51 09 pm](https://user-images.githubusercontent.com/29208142/43474325-8e0d57dc-94c0-11e8-993d-f769249b8f62.png)

Any help would be greatly appreciated. I can provide more information if needed. Thank you!!!

**Comment by amazingjoe** [comment]
You are white-listing/setting remote site settings with https but using http in your browser call. Make the local call is secure over https and I think it should work. Or use http instead of https inside of Salesforce.

Note that Salesforce may not be permitting http even for localhost. If they do allow for http whitelisting for localhost that would probably be the only exception allowed as they typically don't permit this.

**Comment by scottmcconnell5** [comment]
I tried with a publicly hosted URL and it shows the browser call with https, still nothing :(

![screen shot 2018-07-31 at 4 53 13 pm](https://user-images.githubusercontent.com/29208142/43486609-462d77fe-94e2-11e8-8f9c-73c2f2ff5637.png)

**Comment by amazingjoe** [comment]
Can you copy and paste everything in that last portion, some of it is truncated?

Also I am assuming in SF that you have opened access for nthround.io.

**Comment by scottmcconnell5** [comment]
This screenshot shows that entire error, this is the error I get when I turn on my CORS enabling chrome extension... 

![screen shot 2018-08-01 at 12 00 07 pm](https://user-images.githubusercontent.com/29208142/43533408-be6e4fc2-9582-11e8-9080-f2e88822aa3d.png)

**Comment by amazingjoe** [comment]
I think you might have just hit a F-cked up SF technical brick wall.

Take a look at the last post in this thread.

https://developer.salesforce.com/forums/?id=9060G000000XfhRQAS

Essentially you can set up your instance to whitelist the different domains etc... but the authentication server is not your server and not under your purview to whitelist on behalf of.  Which sucks.

It looks like even people using Lighting Out (Salesforce Lighting Components served outside of Salesforce in Javascript) have to find a means around this to get a token.

Here is an example of one proxy that does so. I am betting there are probably prebuilt one button Heroku proxies you can find that will authenticate for you and return the token.

This link has some information on how to get authenticated but it involves setting up nodejs etc.... to do so. http://cloudyworlds.blogspot.com/2016/03/taking-your-lightning-components.html

So essentially it looks like you need to use a proxy to get your tokens and leverage CORS to avoid using a proxy. 

https://stackoverflow.com/questions/29670703/how-to-use-cors-anywhere-to-reverse-proxy-and-add-cors-headers

Now since you probably don't want to do all that work I think you could leverage a CORS Proxy to perform the authentication and get your token and then after that you should be good to go *hopefully*. You will need to set up the authentication for the remote app and whitelists for the proxy server I am thinking.

**Comment by scottmcconnell5** [comment]
@amazingjoe Thank you so much for all this information!! I've spent some time today trying to do what you described, but I still haven't been able to get it running. I followed the instructions laid out in your third link, but I still have a few questions... Do I need to set up my app with Heroku to do this? Or can I just use a CORS proxy to perform the authentication? If I can just use a proxy, I found a promising looking one [here](https://nofluffjuststuff.com/blog/james_ward2/2014/06/cross_origin_resource_sharing_cors_for_salesforce_com) but I've never used a proxy so I'm having trouble. It asks for my sessionID, but how do I get that?

**Comment by amazingjoe** [comment]
Take a look at this project: https://github.com/jsforce/jsforce-ajax-proxy

The ideal would be to get this installed on a cloud provider such as Heroku and then you probably don't have to fiddle with the CORS settings at all if you pipe all your requests through it.

Or you can fiddle with the CORS once you retrieve your session ID which you would get after a successful OAuth.

There are probably easier proxies to install that have these Deploy to Heroku buttons where if you are logged in it pretty much does everything for you. I wasn't able to find one on a quick search but I do imagine they are out there.

Let me know if this helps.

**Comment by scottmcconnell5** [comment]
So I found what looks to be an easy-to-install proxy with a Deploy To Heroku button [here](https://elements.heroku.com/buttons/3vot/3vot-salesforce-proxy), but I'm getting confused with which URL's to put where. If my app is hosted at nthroud.io and I want to be able to connect to the proxy through my app, how should I set my Connected App start url and callback uri? The directions just say to put "xxxx.herokuapp.com" but I'm not sure what to put in for the x's.

Also, when I deploy that proxy to Heroku and launch the app, it brings me to a page with a login hyperlink to login.salesforce.com/xxx
but this link just shows: "error=redirect_uri_mismatch&error_description=redirect_uri%20must%20match%20configuration"

Seems to be an issue with my redirect uri... Any idea what I should be setting all these URL's and uri's to?
